rust/connlib/tunnel/src/device_channel/tun_windows.rs

```diff
@@ -336,14 +336,209 @@ fn set_iface_config(luid: wintun::NET_LUID_LH, mtu: u32) -> Result<()> {
     Ok(())
 }
 
+#[cfg(test)]
 mod tests {
-    /// Checks for regressions in issue #4765, un-initializing Wintun
+    use super::Tun;
+    use anyhow::Result;
+    use ip_packet::{
+        udp::MutableUdpPacket, IpPacket, MutableIpPacket, MutablePacket as _, Packet as _,
+        PacketSize as _,
+    };
+    use std::{
+        future::poll_fn,
+        net::{Ipv4Addr, Ipv6Addr},
+        time::Duration,
+    };
+    use tokio::{
+        net::UdpSocket,
+        time::{timeout, Instant},
+    };
+    use tracing_subscriber::EnvFilter;
+
+    /// Runs multiple Wintun tests.
+    ///
+    /// We don't have test names and UUIDs for the interfaces, and some are performance
+    /// tests, so we don't want to run these in parallel in multiple tests.
     #[test]
     #[ignore = "Needs admin privileges"]
-    fn resource_management() {
+    fn wintun() {
+        let _ = tracing_subscriber::fmt()
+            .with_env_filter(EnvFilter::from_default_env())
+            .with_test_writer()
+            .try_init();
+        perf().unwrap();
+        tunnel_drop();
+    }
+
+    /// Synthetic performance test
+    ///
+    /// Echoes UDP packets between a local socket and the Wintun interface
+    fn perf() -> Result<()> {
```

Comment:
I think it wouldn't be very difficult to extract a small binary in `firezone-tunnel` that just starts a `Tun` device and its only job is to echo packets back that it received.

Once we have that, we can also write a small client CLI that opens a UDP socket and sends them to a certain address, expecting it to be echoed. It can do that for 10s and then exit with the number of bytes it received on stdout.

Bounding the time is more relatable IMO than bounding the number of requests you want to send.

The actual assertion of how many bytes we send can then happen in the script that calls it.

Reply:
Alternatively, we could also try to use https://nikolaivazquez.com/blog/divan/ or criterion to write a benchmark.

Reply:
I'd love to have diagnostic binaries like that, it's just in the past I've gotten pushback about leaving diagnostics and tests in the repo

I prefer subcommands over small binaries because Rust is so slow to build and link, and usually (though I haven't profiled it lately) if all the dependencies are shared, it's faster to build one big binary than multiple small ones. So I'd either like them in the production binary or at least one big binary with all the diagnostics in it.

> The actual assertion of how many bytes we send can then happen in the script that calls it.

I wanted to ask about that, too. I really dislike Bash. This will come up when I rewrite the smoke tests for the Tauri Client. We could switch to Python, I know some Python, but I also dislike it. Node.js is there, but I don't know it well so I'll be slow as I learn it. I'd really rather just write everything in Rust if the tests won't change often. And part of the smoke test is already in Rust since it has to push buttons inside our code that aren't exposed outside the process

Reply:
> I'd love to have diagnostic binaries like that, it's just in the past I've gotten pushback about leaving diagnostics and tests in the repo
> 
> I prefer subcommands over small binaries because Rust is so slow to build and link, and usually (though I haven't profiled it lately) if all the dependencies are shared, it's faster to build one big binary than multiple small ones. So I'd either like them in the production binary or at least one big binary with all the diagnostics in it.

Where would you put it? A subcommand in `firezone-headless-client`?

The linking time is annoying, I agree. The thing I am worried about with bundling too many things together is that they become dependency magnets. Both in terms of cargo dependencies and module dependencies. If they are conceptually unrelated, having them separate means we can more easily move things around and thus iterate on the code faster.

For example, I'd actually like the various `Tun` implementations to move into the platform-specific crate like `connlib-client-android` etc. IMO that is where they belong conceptually and it would make platform-dependent code easier.

Perhaps this wouldn't be impacted by an additional dependency on `Tun` but I hope it still conveys the idea! If we have dedicated binaries for the Windows benchmark, making a decision on where they need to go with such a refactoring is easier.

Whilst I dislike long compile times / link times, I even more dislike convoluted dependencies that block me in writing the code I'd like to write :)

I am open to the repository housing diagnostics and other useful tools. I am also not against expanding the scope of the prod binaries. Esp. debugging tools are something that our customers might also need!

An IP echo server though sounds like something we shouldn't ship.

> > The actual assertion of how many bytes we send can then happen in the script that calls it.
> 
> I wanted to ask about that, too. I really dislike Bash. This will come up when I rewrite the smoke tests for the Tauri Client. We could switch to Python, I know some Python, but I also dislike it. Node.js is there, but I don't know it well so I'll be slow as I learn it. I'd really rather just write everything in Rust if the tests won't change often. And part of the smoke test is already in Rust since it has to push buttons inside our code that aren't exposed outside the process

I don't like bash either. What do you think of tailoring the CLI to take a parameter for minimum transfer speed and it fails with exitcode 1 if that is not reached?

Reply:
> I don't like bash either. What do you think of tailoring the CLI to take a parameter for minimum transfer speed and it fails with exitcode 1 if that is not reached?

That sounds good. We already have `clap` (although it slows down compiles a bit I think) and that would let us adjust the speed without recompiling

> An IP echo server though sounds like something we shouldn't ship.

Agreed, that doesn't need to be on customer's computers. If it's a diagnostic exe then I can still copy it to the resource-constrained VM to see how it runs with 1 CPU core. It's possible to do that with the `cargo test` harness but the exe is hard to find in `target/`

So to recap everything:
- We need to pull this out of the unit tests because (presumably) it needs to build with `--release`, and the unit tests can't do that
- If we're extracting this, it can probably fit into some architecture that the smoke tests use too

However, the smoke tests will be just scripts, aside from the hooks compiled into the GUI for things like simulating the "Export logs" button.

So I agree that multiple binaries would be more convenient to start, the smoke test binary can be unoptimized, and the benchmark binary (Or we could use whatever benchmark Cargo has built in?) will be `--release`.

There is some cargo add-on for running "scripts" too, I should check on that before starting on all this next week. Just so I'm not doing duplicate work.

Also, I'd like to keep the "Send X number of packets and measure the time" instead of "Send packets until Y time has passed" because then I only need to do `Instant::now` twice instead of on every packet, or every batch of packets. I don't have hard numbers, but reading the time must be more expensive than a normal for-loop.

Reply:
> Also, I'd like to keep the "Send X number of packets and measure the time" instead of "Send packets until Y time has passed" because then I only need to do `Instant::now` twice instead of on every packet, or every batch of packets. I don't have hard numbers, but reading the time must be more expensive than a normal for-loop.

The server doesn't need to keep the time at all and the client can have two tasks:

- one that sleeps for 10s
- one that sends / receives UDP traffic

We did some perf work at libp2p a year ago that was inspired by QUIC perf stuff. See https://github.com/libp2p/specs/blob/master/perf/perf.md.

Might be a good inspiration for something here :)

The easiest to implement is likely a download because then the client can easily keep the state for how many packets received and we don't need to invent another control protocol on the side. And we can assume that download and upload are similar through wintun I guess?

Just doing download seems like a good scope that we can get a quick benefit from without pouring too much time into this. What do you think?

Reply:
So like the epoch stopping in wasm runtimes. It would work, I'd still divide by the time just to be safe.
The reason I was testing upload is because the worker thread only handles uploads, and I recently had 2 or 3 bugs in there so I wanted it covered. We could do download too as long as we're here I guess?

Reply:
Yeah thinking about it more, I think we will need some kind of control protocol anyway unless we want to hardcode everything.

By that I mean, i.e. all UDP messages on port 9999 are control messages and we tell it something like: "send us X bytes on port Y" and then we can measure the time it takes to receive those.

Similarly for upload, we'd be sending bytes until the server tells us it us done on another port?

Reply:
That's only needed if the client and server are in separate processes or something, right?

Reply:
Yeah! I was thinking of splitting it but that is probably unnecessary for now!